src/lexer.rs

```diff
@@ -365,6 +388,62 @@ impl<'a> Lexer<'a> {
 
         Some(self.text_token(start, TokenKind::Integer))
     }
+
+    fn read_decimal_part(&mut self, start: usize) -> Option<Token<'a>> {
+        self.read_char(); // consume the '.'
+
+        while self.peek_char().is_ascii_digit() {
+            self.read_char();
+        }
+
+        let p = self.peek_char();
+        match p {
+            'e' => self.read_exponent(start),
+            'E' => self.read_exponent(start),
+            _ => self.read_number_final_floating_point(start),
+        }
+    }
+
+    fn read_exponent(&mut self, start: usize) -> Option<Token<'a>> {
+        self.read_char(); // consume the 'e' or 'E'
+
+        let p = self.peek_char();
+        if p == '+' || p == '-' {
+            self.read_char();
+        }
+        else if !p.is_ascii_digit()
+        {
+            return self.read_number_cleanup_junk(start);
+        }
+
+        self.read_number_final_floating_point(start)
+    }
+
+    fn read_number_final_floating_point(&mut self, start: usize) -> Option<Token<'a>> {
+        while self.peek_char().is_ascii_digit() {
+            self.read_char()
+        }
+
+        let p = self.peek_char();
+        if p.is_ascii_alphabetic() {
+            return self.read_number_cleanup_junk(start);
+        }
+        
+        Some(self.text_token(start, TokenKind::FloatingPoint))
+    }
+
+    fn read_number_cleanup_junk(&mut self, start: usize) -> Option<Token<'a>> {
+        // What we have read is not compatible with a number.  We cannot
+        // rely on the general junk reader to tidy up because we may have
+        // decimal points, exponent symbols and sign symbols in the mix.
+
+        while self.peek_char().is_ascii_alphanumeric() {
```

Comment:
I'd be tempted to consume more characters and group everything that could be part of a malformed number into a single token. (ie include '.', 'e', 'E' '+', '-').

The logic as I understand it would turn 1.23e-4.56 into lots of tokens and record each as unknown. It's probably best if the text of an unknown token is visually nonsensical rather than nonsensical only in context.

I don't feel strongly and we can re-visit this later.

Reply:
For the example you gave, `1.23e-4.56`, it currently produces
```
("1.23e-4", TokenKind::FloatingPoint),
(".56", TokenKind::Unknown)
```

I think it does make sense to consume all characters that could be part of a number. My original concern was something like

`1.23e-4.5+2.6` which should be tokenised as `1.23e-4.5 (Invalid)`,  `+ (Plus)`, `2.6 (FloatingPoint)`, and not as `1.23e-4.5+2.6 (Invalid)`. But perhaps it doesn't matter that there was a more restricted invalid token, as long as we can report the error.